drivers/can/can_sam.c

```diff
@@ -0,0 +1,239 @@
+			const struct zcan_filter *filter)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	return can_mcan_attach_isr(mcan_data, msg_ram, isr, cb_arg, filter);
+}
+
+void can_sam_detach(struct device *dev, int filter_nr)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	can_mcan_detach(mcan_data, msg_ram, filter_nr);
+}
+
+int can_sam_runtime_configure(struct device *dev, enum can_mode mode,
+			       struct can_timing *timing,
+			       struct can_timing *timing_data)
+{
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+
+	return can_mcan_runtime_configure(mcan_cfg, mode, timing, timing_data);
+}
+
+void can_sam_line_0_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	can_mcan_line_0_isr(mcan_cfg, msg_ram, mcan_data);
+}
+
+void can_sam_line_1_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	can_mcan_line_1_isr(mcan_cfg, msg_ram, mcan_data);
+}
+
+static const struct can_driver_api can_api_funcs = {
+	.configure = can_sam_runtime_configure,
+	.send = can_sam_send,
+	.attach_isr = can_sam_attach_isr,
+	.detach = can_sam_detach,
+	.get_state = can_sam_get_state,
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	.recover = can_mcan_recover,
+#endif
+	.can_get_core_clock = can_sam_get_core_clock,
+	.register_state_change_isr = can_sam_register_state_change_isr,
+	.timing_min = {
+		.sjw = 0x7f,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x01,
+		.phase_seg2 = 0x01,
+		.prescaler = 0x01
+	},
+	.timing_max = {
+		.sjw = 0x7f,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x100,
+		.phase_seg2 = 0x80,
+		.prescaler = 0x200
+	},
+#ifdef CONFIG_CAN_FD_MODE
+	.timing_min_data = {
+		.sjw = 0x01,
+		.prop_seg = 0x01,
+		.phase_seg1 = 0x01,
+		.phase_seg2 = 0x01,
+		.prescaler = 0x01
+	},
+	.timing_max_data = {
+		.sjw = 0x10,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x20,
+		.phase_seg2 = 0x10,
+		.prescaler = 0x20
+	}
+#endif
+};
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(can0), okay)
+
+static void config_can_0_irq();
+
+static const struct can_sam_config can_sam_cfg_0 = {
+	.pmc_id = DT_PROP(DT_NODELABEL(can0), peripheral_id),
+	/*.pin_list = {
+		ATMEL_SAM_DT_PIN(0, 0),
+		ATMEL_SAM_DT_PIN(0, 1)
+	},*/
```

Comment:
pin list need listfy, you need use:
`.pin_list = ATMEL_SAM_DT_PINS(n),`
You can need to:
`.num_pins = ATMEL_SAM_DT_NUM_PINS(n),`
Take look at drivers/spi/spi_sam.c
The definition is here:
./soc/arm/atmel_sam/common/atmel_sam_dt.h
The idea is all Atmel DT macros should be place there to easy reuse.